core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/CompactExpressionEvaluator.scala

```diff
@@ -0,0 +1,382 @@
+        .constructFinalInputValues(
+          combinedResults,
+          parsedTree,
+          Some(Seq(entityName)),
+          None
+        )
+        .getOrElse(entityName, Success(Seq.empty))
+        .get
+        .toSeq
+    }
+  }
+
+  /**
+   * Evaluates multiple expressions against entities in the context of a workflow submission,
+   * handling entity type transitions and optimizing database queries across all expressions.
+   *
+   * This method is designed for workflow submission validation where multiple input expressions
+   * need to be evaluated against potentially different entity types. It handles cases where
+   * the input entity type differs from the target entity type (rootEntityType) by properly
+   * traversing entity relationships. The method optimizes performance by consolidating all
+   * expression lookups into efficient query plans and executing them in batch.
+   *
+   * @param workspaceId The UUID of the workspace containing the entities to evaluate against
+   * @param expressionEvaluationContext Context containing:
+   *                                   - entityType: The type of the input entity (e.g., "sample_set")
+   *                                   - entityName: The name of the input entity
+   *                                   - expression: Optional entity expression to determine target entities
+   *                                   - rootEntityType: The expected entity type for final results (e.g., "sample")
+   * @param gatherInputsResult The workflow inputs that need expression evaluation, containing
+   *                          processableInputs with their expressions
+   * @return A Future containing a LazyList of SubmissionValidationEntityInputs, where each
+   *         entry represents the resolved input values for a specific entity of the rootEntityType
+   *
+   * @throws RawlsExceptionWithErrorReport if:
+   *         - Required context fields (entityType, entityName, rootEntityType) are missing
+   *         - Entity type doesn't match rootEntityType when no entity expression is provided
+   *         - Expressions cannot be parsed or referenced entities/attributes don't exist
+   *
+   */
+  def evaluateExpressions(workspaceId: UUID,
+                          expressionEvaluationContext: ExpressionEvaluationContext,
+                          gatherInputsResult: MethodConfigResolver.GatherInputsResult
+  )(implicit executionContext: ExecutionContext): Future[LazyList[SubmissionValidationEntityInputs]] = {
+    // First, verify that necessary entity information is present and consistent
+    val entityType = expressionEvaluationContext.entityType.getOrElse(
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Missing entityType"))
+    )
+    val entityName = expressionEvaluationContext.entityName.getOrElse(
+      throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Missing entityName"))
+    )
```

Comment:
you could add:
```
    EntityUtils.validateEntityType(entityType)
    EntityUtils.validateEntityName(entityName)
```
for extra measure. I suppose it would help ensure that the expressions are valid.